### PR TITLE
[sym_shape][perf] eval_static: guards, unbacked compute once

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124217

Differential Revision: [D56212345](https://our.internmc.facebook.com/intern/diff/D56212345)

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -3623,8 +3641,8 @@ def add_expr(expr):
                     add_expr(sympy.Le(e.lhs, e.rhs))
                     add_expr(sympy.Ne(e.lhs, e.rhs))
 
-            # NB: this helps us deal with And/Or connectives
-            expr = expr.subs(subst)
+		# NB: this helps us deal with And/Or connectives
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:3644]: tabs

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -3593,6 +3593,27 @@ def _maybe_evaluate_static(
         symbols = list(expr.free_symbols)
 
         # Apply known runtime asserts
+        guards_exprs = []
+        for g in self.guards:
+            e = self.simplify(g.expr)
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:3598]: On master, we're not calling simplify yet. I guess it's harmless enough to do here though.

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -3593,6 +3593,27 @@ def _maybe_evaluate_static(
...
+            if compute_hint:
+                e = canonicalize_bool_expr(e.xreplace(self.var_to_val))
+            guards_exprs.append(e)
+
+        symbols_unbacked = symbols - self.var_to_val.keys()
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:3603]: Isn't this set on symbols, aren't these slow?

**Review by ezyang — APPROVED** [review]

**Comment by IvanKobzarev** [comment]
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56212345).

File: torch/fx/experimental/symbolic_shapes.py (hunk 3 above)
Comment (IvanKobzarev) [reply torch/fx/experimental/symbolic_shapes.py:3603]: Yes, this is similar to that strange optimization, but profile did not show this line

**Comment by IvanKobzarev** [comment]
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56212345).

File: torch/fx/experimental/symbolic_shapes.py (hunk 3 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:3603]: yeah, this is partially why I feel there was something very strange going on with that particular optimization.

**Comment by lezcano** [comment]
Another approach to the same issue: https://github.com/pytorch/pytorch/pull/123342/

**Comment by IvanKobzarev** [comment]
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56212345).

**Pull request closed without merging**